gerritlab/global_vars.py

```diff
@@ -27,28 +34,122 @@ def load_config(remote, repo):
+                       gitreview_config: configparser.SectionProxy):
+    """
+    Get the private token from one of several sources in the following order of
+    preference:
+
+    1. The GITLAB_PRIVATE_TOKEN environment variable.
+    2. git credential storage
+    3. git config
+    3. (DEPRECATED) The .gitreview file
+
+    Kind of a messy function, but it hides this complexity from the
+    `load_config` function.
+    """
+    if "GITLAB_PRIVATE_TOKEN" in os.environ:
+        return os.environ["GITLAB_PRIVATE_TOKEN"]
+
+    # Try to get the private token from git credentials.
+    private_token = GitCredentialStore(host).get_token()
+    if private_token is not None:
+        return private_token
+
+    # Try to get the private token from git config.
+    try:
+        return git_config.get_value("gerritlab", "private-token")
+    except (configparser.NoSectionError, configparser.NoOptionError):
+        pass
+
+    # DEPRECATED: Try to get the private token from .gitreview file.
+    if "private_token" in gitreview_config:
+        warn("[DEPRECATED] Using private_token from .gitreview file is deprecated.")
+        return gitreview_config["private_token"]
+
+    raise SystemExit(f"Unable to find private token for {host}")
+
+
+class GitCredentialStore:
+    """
+    A class to interact with git credential store.
+    """
+    def __init__(self, host:str) -> None:
+        # TODO: use host + repo path to support per-repo tokens
+        self._host = host
+        self._token: Optional[str] = None
+        self._git_credentials: Dict[str,str] = {}
+
+    def get_token(self) -> Union[str, None]:
+        """
+        Convenience method to get the token from the credential store.
+        """
+        return self.get("password")
+
+    def get(self, key: str, default=None) -> Union[str, None]:
+        """
+        Return key from internal git credentials cache, or default if not found.
+        """
+        if self._git_credentials.get("fetched") is None:
+            # Isolate all IO here for ease of testing
+            self._populate_git_credentials(self._call_git_credential_fill())
+
+        return self._git_credentials.get(key, default)
+
+    def _populate_git_credentials(self, git_credentials_output: str) -> None:
+        """
+        Populate the internal git credentials cache
+        """
+        now = datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
+        self._git_credentials["fetched"] = now
+
+        for line in git_credentials_output.splitlines():
+            if "=" in line:
+                key, value = line.split("=", 1)
+                self._git_credentials[key] = value
+
+    def _call_git_credential_fill(self) -> str:
+        """
+        Look up credentials for the host using git credential fill.
+
+        Prepend 'gerritlab@' to ensure gitlab tokens can be stored separately
+        from other user tokens if desired.
+        """
+        git_credentials = subprocess.run(
```

Comment:
I don't have a credentials store configured so when I run `git lab` with this code in place, it ends up prompting me for username/password where it didn't before (because I have the token in my ~/.gitconfig).  Because of this I think checking the credential store should be the last thing that is attempted.

Reply:
Moved to the last option.